### PR TITLE
[ASL] Various features

### PR DESCRIPTION
**Pull request by HadrienRenaud**
Major features
- [x] `.item` syntax to access tuple elements.
- [x] Case guards semantics and type-checking.

Type-checking and other safety features
- [x] Non overlapping slices on left-hand-side expressions.
- [x] Prevent empty enumerations.
- [x] Tuples are parameter-defining.
- [x] Ensure getters have corresponding setters.

And some bug fixes.

### CODE REVIEW

File: asllib/ASTUtils.ml [hunk 1]
```diff
@@ -165,138 +165,138 @@ let slices_to_positions as_int =
 let fold_named_list folder acc list =
   List.fold_left (fun acc (_, v) -> folder acc v) acc list
 
-let rec use_e acc e =
+let ( $ ) f1 f2 acc = f1 acc |> f2
+
+let use_option use_thing = function
```
Comment (jalglave) [inline asllib/ASTUtils.ml:170]: what can this "thing" be here?

File: asllib/ASTUtils.ml [hunk 2]
```diff
@@ -165,138 +165,138 @@ let slices_to_positions as_int =
...
+  | None -> Fun.id
+  | Some thing -> use_thing thing
```
Comment (jalglave) [inline asllib/ASTUtils.ml:172]: same question here - what could thing be?

File: asllib/builder.ml [hunk 3]
```diff
@@ -168,7 +168,7 @@ let stdlib =
   in
   lazy
     (match list_first_opt_opt try_one to_try with
-    | Some ast -> offuscate ast
+    | Some ast -> if false then ast else offuscate ast
```
Comment (jalglave) [inline asllib/builder.ml:171]: typo: obfuscate

File: asllib/tests/regressions.t/bug5.asl [hunk 4]
```diff
@@ -0,0 +1,10 @@
+func test(N: integer, a: bits(N))
```
Comment (jalglave) [inline asllib/tests/regressions.t/bug5.asl:1]: please might you invent more explicit names than bug*.asl?

**Review by jalglave — COMMENTED** [review]
thanks @HadrienRenaud - some very small things.

**Review by Roman-Manevich — COMMENTED** [review]
comments inline

**Pull request merged**